// ng6x-editor/fesm2015/ng6x-editor-projects-ngx-editor.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgxEditorService {
    constructor() { }
}
NgxEditorService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgxEditorService.ctorParameters = () => [];
/** @nocollapse */ NgxEditorService.ngInjectableDef = defineInjectable({ factory: function NgxEditorService_Factory() { return new NgxEditorService(); }, token: NgxEditorService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgxEditorComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
NgxEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-ngx-editor',
                template: `
    <p>
      ngx-editor works!
    </p>
  `
            }] }
];
/** @nocollapse */
NgxEditorComponent.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgxEditorModule {
}
NgxEditorModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgxEditorComponent],
                exports: [NgxEditorComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgxEditorService, NgxEditorComponent, NgxEditorModule };
