// ng6x-editor/bundles/ng6x-editor-projects-ngx-editor.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('ng6x-editor/projects/ngx-editor', ['exports', '@angular/core'], factory) :
    (factory((global['ng6x-editor'] = global['ng6x-editor'] || {}, global['ng6x-editor'].projects = global['ng6x-editor'].projects || {}, global['ng6x-editor'].projects['ngx-editor'] = {}),global.ng.core));
}(this, (function (exports,i0) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxEditorService = (function () {
        function NgxEditorService() {
        }
        NgxEditorService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        NgxEditorService.ctorParameters = function () { return []; };
        /** @nocollapse */ NgxEditorService.ngInjectableDef = i0.defineInjectable({ factory: function NgxEditorService_Factory() { return new NgxEditorService(); }, token: NgxEditorService, providedIn: "root" });
        return NgxEditorService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxEditorComponent = (function () {
        function NgxEditorComponent() {
        }
        /**
         * @return {?}
         */
        NgxEditorComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        NgxEditorComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-ngx-editor',
                        template: "\n    <p>\n      ngx-editor works!\n    </p>\n  "
                    }] }
        ];
        /** @nocollapse */
        NgxEditorComponent.ctorParameters = function () { return []; };
        return NgxEditorComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxEditorModule = (function () {
        function NgxEditorModule() {
        }
        NgxEditorModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [],
                        declarations: [NgxEditorComponent],
                        exports: [NgxEditorComponent]
                    },] }
        ];
        return NgxEditorModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgxEditorService = NgxEditorService;
    exports.NgxEditorComponent = NgxEditorComponent;
    exports.NgxEditorModule = NgxEditorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
