// ng6x-editor/fesm5/ng6x-editor-projects-ngx-editor.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxEditorService = /** @class */ (function () {
    function NgxEditorService() {
    }
    NgxEditorService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    NgxEditorService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxEditorService.ngInjectableDef = defineInjectable({ factory: function NgxEditorService_Factory() { return new NgxEditorService(); }, token: NgxEditorService, providedIn: "root" });
    return NgxEditorService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxEditorComponent = /** @class */ (function () {
    function NgxEditorComponent() {
    }
    /**
     * @return {?}
     */
    NgxEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    NgxEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ngx-editor',
                    template: "\n    <p>\n      ngx-editor works!\n    </p>\n  "
                }] }
    ];
    /** @nocollapse */
    NgxEditorComponent.ctorParameters = function () { return []; };
    return NgxEditorComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxEditorModule = /** @class */ (function () {
    function NgxEditorModule() {
    }
    NgxEditorModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [NgxEditorComponent],
                    exports: [NgxEditorComponent]
                },] }
    ];
    return NgxEditorModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgxEditorService, NgxEditorComponent, NgxEditorModule };
